### ISSUE TITLE
PG - Updating gradient every step correct? (vs summing the gradient?)

### ISSUE THREAD

**Issue by drozzy** [thu-ml/tianshou#404]
Hi, I'm trying to understand why your per-step update in PG is correct:
https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L133

over something like in [pytorch example ](https://github.com/pytorch/examples/blob/cedca7729fef11c91e28099a0e45d7e98d03b66d/reinforcement_learning/reinforce.py#L71)(where they add up all the loss terms and update it just once):
```
for log_prob, R in zip(policy.saved_log_probs, returns):
        policy_loss.append(-log_prob * R)
    optimizer.zero_grad()
    policy_loss = torch.cat(policy_loss).sum()
    policy_loss.backward()
    optimizer.step()
```

I know that per-step update is what is shown in the RL Sutton & Barto book - but it doesn't make sense to me.
It seems off-policy almost!
Because the theta you're updating at all timesteps (except the very first timestep) is no longer the theta that was used for experience collection.

Thanks.

**Comment by Trinkle23897** [comment]
@ChenDRAG

**Comment by drozzy** [comment]
Ok I think I understand. The [`batch_size`](https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L119) defines the subset of the Batch object that will be used in feeding the policy's neural network (for the optimizer step/update).
So if I set the batch size to be equal to the episode length (assuming that was possible - since episode length is dynamic), then I would get the equivalent of the pytorch update.
But if I set `batch_size=1` then I would get an update each step - which would be equivalent to the update used in the RL book.

I find the naming of [batch](https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L120) and [`b`](https://github.com/thu-ml/tianshou/blob/c19876179a52acdc4d2e7dcb10dfbb4557adca86/tianshou/policy/modelfree/pg.py#L124) batch confusing, but I can't think of a better way of naming them. I guess that's ok.

**Comment by Trinkle23897** [comment]
Yes you can set batch-size to max episode length, in this way it will be equivalent to the scenario you described.
